### ISSUE TITLE
32/64-bit difference in mpileup indel caller.

### ISSUE THREAD

**Issue by jkbonfield** [samtools/samtools#305]
I've been fixing some of these, including a buffer overrun, but I cannot identify the problem with this particular test.

On 64-bit systems after a build in samtools (develop branch):
./samtools mpileup -x -F 0.60 -vu -f test/mpileup/mpileup.ref.fa ./test/mpileup/indels.sam|grep INDEL
[mpileup] 2 samples in 1 input files
<mpileup> Set max per-file depth to 4000
##INFO=<ID=INDEL,Number=0,Type=Flag,Description="Indicates that the variant is an INDEL.">
17      547     .       A       AGCATCCTGCG     0       .       INDEL;IDV=2;IMF=0.666667;DP=5;I16=2,0,1,2,404,81608,538,130834,120,7200,180,10800,38,722,63,1347;QS=1.01515,0.984848;VDB=0.187998;SGB=-0.8159;MQSB=1;MQ0F=0       PL      111,0,51        22,0,54

On 32-bit system (i386) it doesn't find the indel:
./samtools mpileup -x -F 0.60 -vu -f test/mpileup/mpileup.ref.fa ./test/mpileup/indels.sam|grep INDEL
[mpileup] 2 samples in 1 input files
<mpileup> Set max per-file depth to 4000
##INFO=<ID=INDEL,Number=0,Type=Flag,Description="Indicates that the variant is an INDEL.">

There is nothing reported by Valgrind. I asume it is some type overflow condition.  All data is within the samtools test suite.

**Comment by jmarshall** [comment]
Alternatively to the third paragraph of my previous comment, another fix is to declare `min_frac` as **long double** on i386.  Then the comparison will work as expected on this platform.  Ideally it'd stay as **float** on other platforms though.

**Comment by jkbonfield** [comment]
Oops, I even managed to submit a duplicate of *my own ticket*.

Still, it needs fixing or else if we deem it to be such a marginal case then the test needs changing to not fail. (Ideally the former!)

**Comment by jkbonfield** [comment]
Doesn't long double then open questions as to how you do an atof on a string to get the long double in the first place? (See "mplp.min_frac = atof(optarg);" in bam_plcmd.c.

However If that does work, one solution may be just to use 64-bit integer arithmetic.  If we restrict n_alt and n_tot to 32-bit (they are already) then n_alt / n_tot < min_frac is the same as n_alt < n_tot * min_frac.

If min_frac is actually the floating point bcf->min_frac << 32 then we can do (((int64_t)n_alt) << 32) < n_tot * min_frac.... I think!  The entire thing then comes integer arithmetic except for the initial shift.

**Comment by jkbonfield** [comment]
Indeed...

```
#include <stdlib.h>
#include <stdio.h>
#include <inttypes.h>

int main(int argc, char **argv) {
    int a = atoi(argv[1]);
    int n = atoi(argv[2]);
    //float min_frac = n*atof(argv[3]);
    float min_frac = atof(argv[3]);
    int64_t min_frac64 = min_frac * (((int64_t)1)<<32);

    if ((float)a / n < min_frac)
        puts("<");
    else
        puts(">=");

    if ((float)a < n*min_frac)
        puts("<");
    else
        puts(">=");

    if ((double)a / n < min_frac)
        puts("<");
    else
        puts(">=");

    if ((float)a < n*(double)min_frac)
        puts("<");
    else
        puts(">=");

    if ((((int64_t)a)<<32) < n*min_frac64)
        puts("<");
    else
        puts(">=");

    return 0;
}

// 3 5 0.5999999940395354669499999
// 3 5 0.5999999940395354669500000
```

With the two fractions listed in the comments at the end, on 64-bit we get the same difference that the 32-bit version does with 3 of the 5 variants.

Whether this is just barking up the wrong tree I'm unsure.  Possibly the problem is simply the test being too strict, but 0.6 is a nice round number that is likely to be used by users, so all we could do with be add some woolly FLT_EPSILON in there which simply moves the problem slightly to a less "nice" number.

**Comment by jkbonfield** [comment]
if ((float)a / n - min_frac < FLT_EPSILON)
        puts("<");
    else
        puts(">=");

is also a stable comparison between 32-bit and 64-bit it appears (0.5999999344348908025899999/900000).  Possibly this is the "correct" comparison.  Certainly it's good that it moves the boundary point of oddity from nice numbers like 0.6 to rather more wordy numbers, which is about as good as you can hope for in fp.

**Comment by jkbonfield** [comment]
I got fed up of it hanging around so went with a FLT_EPSILON workaround.

Someone please review it.

**Comment by jmarshall** [comment]
Somewhat disappointingly (I had `hts_rational` structs and routines all ready to go!), it turns out that just doing the arithmetic as `double` makes x86 work similarly to x86-64 and others.

Maybe this trumps all our other more complicated solutions? :smile:

**Comment by charles-plessy** [comment]
Hi James,

are you sure the problem is solved ?  In Debian's build farm, the regression tests failed with samtools 1.2, on the same mpileup command…

<https://buildd.debian.org/status/fetch.php?pkg=samtools&arch=i386&ver=1.2-1&stamp=1434409216>

```
cd test/mpileup && ./regression.sh
Samtools mpileup tests:

EXPECTED FAIL: Task failed, but expected to fail;
when running $samtools mpileup -x -d 8500 -B -f mpileup.ref.fa deep.sam|awk '{print $4}'
UNEXPECTED FAIL: Output mismatch for $samtools mpileup -x -F 0.60 -u -f mpileup.ref.fa indels.bam|$filter|awk '/INDEL/'
See FAIL-59.out.1 vs expected/59.out

UNEXPECTED FAIL: Output mismatch for $samtools mpileup -x -F 0.60 -u -f mpileup.ref.fa indels.cram|$filter|awk '/INDEL/'
See FAIL-59.out.2 vs expected/59.out


Expected   passes:   121
Unexpected passes:   0
Expected   failures: 1
Unexpected failures: 2
=> FAIL
```

Cheers,

-- 
Charles

**Comment by jkbonfield** [comment]
Samtools 1.2 was released in February, so this fix postdates the 1.2 release.  It's natural to expect it to keep failing until we release 1.3, or have Debian being doing backporting of specific bug fixes?

**Comment by charles-plessy** [comment]
Ah, sorry, I did not pay attention that the commit did not reach yet the master branch.  Thanks for the quick answer !
